pytorch/pytorch_edgeml/trainer/bonsaiTrainer.py

```diff
@@ -0,0 +1,417 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license.
+
+import torch
+import numpy as np
+import os
+import sys
+import pytorch_edgeml.utils as utils
+
+
+class BonsaiTrainer:
+
+    def __init__(self, bonsaiObj, lW, lT, lV, lZ, sW, sT, sV, sZ,
+                 learningRate, useMCHLoss=False, outFile=None, device=None):
+        '''
+        bonsaiObj - Initialised Bonsai Object and Graph
+        lW, lT, lV and lZ are regularisers to Bonsai Params
+        sW, sT, sV and sZ are sparsity factors to Bonsai Params
+        learningRate - learningRate for optimizer
+        useMCHLoss - For choice between HingeLoss vs CrossEntropy
+        useMCHLoss - True - MultiClass - multiClassHingeLoss
+        useMCHLoss - False - MultiClass - crossEntropyLoss
+        '''
+
+        self.bonsaiObj = bonsaiObj
+
+        self.lW = lW
+        self.lV = lV
+        self.lT = lT
+        self.lZ = lZ
+
+        self.sW = sW
+        self.sV = sV
+        self.sT = sT
+        self.sZ = sZ
+
+        if device is None:
```

Comment:
You can probably move this to `train` routine ? That makes more sense ? So that user can switch between `cpu` and `gpu` over different calls to `train` ? 

Also will help with bench-marking i think and debugging.

Reply:
In fact if you think about it, because of the way pytorch works, all the arguments provided to `__init__` can be moved to `train` method and can be dynamically changed.